File: testdata/precision-suite-perl/clean_safe_db.pl
#!/usr/bin/perl
# Clean: the SAFE counterparts of every tp_*.pl flow. None of these should fire —
# each routes the tainted value through the sanitizer / parameterized form that
# neutralizes its sink's vuln class. A finding on any line here is a false
# positive.
use strict;
use warnings;
use DBI;
use File::Basename qw(basename);

# Parameterized DBI: the tainted value is a bind parameter (passed to `do` after
# the SQL string), not interpolated into it — safe against SQL injection.
sub by_user {
    my ($dbh) = @_;
    my $id = $ENV{USER_ID};
    $dbh->do("SELECT * FROM users WHERE id = ?", undef, $id);
}

# Prepared statement with a placeholder, then bind via execute — safe.
sub lookup_name {
    my ($dbh) = @_;
    my $name = $ENV{NAME};
    my $sth  = $dbh->prepare("SELECT * FROM users WHERE name = ?");
    $sth->execute($name);
}

# Integer coercion strips every injection metacharacter before the shell command.
sub ping {
    my $raw   = $ENV{COUNT};
    my $count = int($raw);
    system "ping -c $count example.com";
}

# quotemeta escapes shell metacharacters before the command runs.
sub lookup_host {
    my $host = $ENV{HOST};
    my $safe = quotemeta($host);
    system "host $safe";
}

# File::Basename::basename strips directory components, defusing path traversal.
sub download {
    my $name = $ENV{FILE};
    my $base = basename($name);
    open(my $fh, "<", "/srv/downloads/$base");
    return <$fh>;
}

# No source: a constant command is never tainted.
sub status {
    system "systemctl status nginx";
}

1;

File: testdata/precision-suite-ruby/clean_safe_db.rb
# Clean: the SAFE counterparts of every tp_*.rb flow. None of these should fire —
# each routes the tainted value through the sanitizer / parameterized form that
# neutralizes its sink's vuln class. A finding on any line here is a false
# positive.
class SafeController
  # Parameterized ActiveRecord: the tainted value is a bind parameter (2nd arg),
  # not interpolated into the SQL string — safe against SQL injection.
  def by_user
    id = params[:user_id]
    User.where("id = ?", id)
  end

  # Integer coercion strips every injection metacharacter before the value is
  # used in a shell command.
  def ping
    raw = params[:count]
    count = Integer(raw)
    system "ping -c #{count} example.com"
  end

  # Shellwords.escape neutralizes command injection.
  def lookup
    host = params[:host]
    safe = Shellwords.escape(host)
    system "host #{safe}"
  end

  # File.basename strips directory components, defusing path traversal.
  def download
    name = params[:file]
    base = File.basename(name)
    File.read("/srv/downloads/#{base}")
  end

  # YAML.safe_load refuses to instantiate arbitrary objects — safe deserialization.
  def import
    doc = params[:doc]
    YAML.safe_load(doc)
  end

  # CGI.escapeHTML neutralizes XSS before the value is marked html_safe.
  def preview
    body = params[:body]
    escaped = CGI.escapeHTML(body)
    @html = escaped.html_safe
  end

  # No source: a constant command is never tainted.
  def status
    system "systemctl status nginx"
  end
end
